### PR TITLE
Command Refactoring

### PR DESCRIPTION
**Pull request by grolu**
**What this PR does / why we need it**:
Refactor commands
- Flatten hierarchy, to have command and all subcommands in one package
- Have test and code in same package (like kubectl)
- Combine options, completions and command implementation in one file (like kubectl)

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```feature user

```

### CODE REVIEW

File: pkg/cmd/root/command.go [hunk 1]
```diff
@@ -53,7 +53,7 @@ func Execute() {
 	}
 
 	rootCmd.AddCommand(ssh.NewCommand(&factory, ssh.NewOptions(ioStreams)))
-	rootCmd.AddCommand(targetcmd.NewCommand(&factory, targetcmd.NewOptions(ioStreams), targetProvider))
+	rootCmd.AddCommand(targetcmd.NewTargetCommand(&factory, targetcmd.NewTargetOptions(ioStreams), targetProvider))
```
Comment (holgerkoser) [inline pkg/cmd/root/command.go:56]:
```suggestion
	rootCmd.AddCommand(target.NewCmdTarget(&factory, ioStreams, targetProvider))
```

Comment (holgerkoser) [reply pkg/cmd/root/command.go:56]: The naming for the function which creates the command is `NewCmd...`. The options are created inside this function. First parameter is the factory, second parameter is the ioStreams object. They do not rename the import of the subcommand package.

File: pkg/cmd/target/drop.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
+/*
```
Comment (holgerkoser) [inline pkg/cmd/target/drop.go:1]: The name of this file should be `target_drop.go`

File: pkg/cmd/target/drop.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+func NewDropOptions(ioStreams util.IOStreams) *DropOptions {
+	return &DropOptions{
+		Options: base.Options{
+			IOStreams: ioStreams,
+		},
+	}
+}
+
+// Complete adapts from the command line args to the data required.
+func (o *DropOptions) Complete(f util.Factory, cmd *cobra.Command, args []string, targetProvider *target.DynamicTargetProvider) error {
```
Comment (holgerkoser) [inline pkg/cmd/target/drop.go:95]:
```suggestion
func (o *DropOptions) Complete(f util.Factory, cmd *cobra.Command, args []string) error {
```

Comment (holgerkoser) [reply pkg/cmd/target/drop.go:95]: targetProvider is not used in this function.

File: pkg/cmd/target/drop.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+		Use:   "drop",
+		Short: "Drop target, e.g. \"gardenctl target drop shoot\" to drop currently targeted shoot",
+		ValidArgs: []string{
+			string(TargetKindGarden),
+			string(TargetKindProject),
+			string(TargetKindSeed),
+			string(TargetKindShoot),
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := o.Complete(f, cmd, args, targetProvider); err != nil {
```
Comment (holgerkoser) [inline pkg/cmd/target/drop.go:33]:
```suggestion
func NewDropCommand(f util.Factory, o *DropOptions) *cobra.Command {
	cmd := &cobra.Command{
		Use:   "drop",
		Short: "Drop target, e.g. \"gardenctl target drop shoot\" to drop currently targeted shoot",
		ValidArgs: []string{
			string(TargetKindGarden),
			string(TargetKindProject),
			string(TargetKindSeed),
			string(TargetKindShoot),
		},
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := o.Complete(f, cmd, args); err != nil {
```

File: pkg/cmd/target/drop.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+			}
+
+			return runDropCommand(f, o)
+		},
+	}
+
+	return cmd
+}
+
+func runDropCommand(f util.Factory, o *DropOptions) error {
```
Comment (holgerkoser) [inline pkg/cmd/target/drop.go:47]:
```suggestion
func runCmdDrop(f util.Factory, o *DropOptions) error {
```

File: pkg/cmd/target/drop.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := o.Complete(f, cmd, args, targetProvider); err != nil {
+				return fmt.Errorf("failed to complete command options: %w", err)
+			}
+			if err := o.Validate(); err != nil {
+				return err
+			}
+
+			return runDropCommand(f, o)
```
Comment (holgerkoser) [inline pkg/cmd/target/drop.go:40]:
```suggestion
			return runCmdDrop(f, o)
```

File: pkg/cmd/target/target.go [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+	"strings"
+
+	"github.com/gardener/gardenctl-v2/internal/util"
+	"github.com/gardener/gardenctl-v2/pkg/target"
+
+	"github.com/spf13/cobra"
+)
+
+// NewTargetCommand returns a new target command.
+func NewTargetCommand(f util.Factory, o *TargetOptions, targetProvider *target.DynamicTargetProvider) *cobra.Command {
```
Comment (holgerkoser) [inline pkg/cmd/target/target.go:22]:
```suggestion
func NewCmdTarget(f util.Factory, o *TargetOptions, targetProvider *target.DynamicTargetProvider) *cobra.Command {
```

File: pkg/cmd/target/drop_test.go [hunk 8]
```diff
@@ -211,3 +210,22 @@ var _ = Describe("Command", func() {
...
+		o := NewDropOptions(streams)
+		o.Kind = TargetKindGarden
+
+		Expect(o.Validate()).To(Succeed())
+	})
+
+	It("should reject invalid kinds", func() {
+		streams, _, _, _ := util.NewTestIOStreams()
+		o := NewDropOptions(streams)
+		o.Kind = TargetKind("not a kind")
```
Comment (petersutter) [inline pkg/cmd/target/drop_test.go:226]: Redundant type conversion

File: pkg/cmd/target/drop_test.go [hunk 9]
```diff
@@ -4,12 +4,11 @@ SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener con
 SPDX-License-Identifier: Apache-2.0
 */
 
-package drop_test
```
Comment (petersutter) [inline pkg/cmd/target/drop_test.go:7]: why? leave it in the `drop_test` package and import cmd target as 
`cmdtarget "github.com/gardener/gardenctl-v2/pkg/cmd/target"`
same applies for the other test package renamings

Comment (grolu) [reply pkg/cmd/target/drop_test.go:7]: sure, we could do so. But at least for the targe_test file, we nee dto have it in the same packe or alternatively we nee to export `validTargetArgsFunction`
So I think wee need to agree on one approach here. Either we always include tests in the same package or we don't but accept that we need to export all things we'd like to test then. WDYT?

Comment (grolu) [reply pkg/cmd/target/drop_test.go:7]: added whitebox test file as discussed

**Review by petersutter — APPROVED** [review]
/lgtm

**Review by holgerkoser — APPROVED** [review]
/lgtm

**Pull request merged**